logs: T3774: Added CLI options to control atop logs rotation

zdc opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Added CLI options to generate logrotate configuration file for atop logs

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T3774

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
logs, logrotate, atop

## Proposed changes
<!--- Describe your changes in detail -->
To give users control over logs we should provide a CLI section for this. The commit adds the section with the first option that can be controlled - atop log file rotation strategy. In the same way, we can add more (`/var/log/messages` is the first and obvious candidate).

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
To test:
```
set system logs logrotate atop maxsize 15
set system logs logrotate atop rotate 25
commit
```
Then check values:
```
vyos@vyos# grep -E 'maxsize|rotate' /etc/logrotate.d/vyos-atop 
    maxsize 15M
    rotate 25
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

zdc committed:
logs: T3774: Added CLI options to control atop logs rotation

Added CLI options to generate logrotate configuration file for atop logs

c-po commented on interface-definitions/system-logs.xml.in:
```diff
@@ -0,0 +1,57 @@
+<interfaceDefinition>
+  <node name="system">
+    <children>
+      <node name="logs" owner="${vyos_conf_scripts_dir}/system-logs.py">
```
Can we name this `system logging` instead?

zdc replied:
I have no predisposition for any of the name variants in this case - this is only the question of taste. Would you like to change the name?

sever-sever replied:
As for me, it logically should be in `set system syslog` section
As we have such description:
```
set system 
...
 > syslog       System logging
...
```
It's just my opinion. @andamasov doesn't like this idea

zdc replied:
@sever-sever, I see at least two different components that are not related anyhow to Syslog and should be controlled by this:
- logrotate
- systemd-journald
That is why it was placed in an independent node.

c-po replied:
I'd also not place it under syslog node as this is logrotate and not syslog ;)

sever-sever replied:
I prefer to set all log options  under one "logical node", which will include all other logs/logging/syslog
I understand that it 2 different functions
But it looks strange
```
set system logging
set system logs
set system syslog (description System logging)
```

c-po replied:
For 1.4 we could also migrate syslog to a child of „system logging“

c-po commented on interface-definitions/system-logs.xml.in:
```diff
@@ -0,0 +1,57 @@
+            <children>
+              <node name="atop">
+                <properties>
+                  <help>Atop logs options</help>
```
What is atop? Maybe explain a bit in the help string.

zdc replied:
Good point. Added short description.

c-po commented on interface-definitions/system-logs.xml.in:
```diff
@@ -0,0 +1,57 @@
+                  <help>Atop logs options</help>
+                </properties>
+                <children>
+                  <leafNode name="maxsize">
```
Please use `max-size` to match other CLI nodes that already exist using this name.

zdc replied:
This is a controversial question. We are controlling here the `logrotate` option with the same name. As for me, I see here no relation to the rest places in CLI, but a clear interface for people who are already familiar with `logrotate`.
In general, sometimes renaming options in our own style is a great idea - for example when we integrate new tools into existing services with already established terminology. I just think that this is not one of such cases.

c-po replied:
It's only a hyphen which makes it easier readable on the CLI ;)

zdc replied:
Replaced to `max-size`.

c-po commented on interface-definitions/system-logs.xml.in:
```diff
@@ -0,0 +1,57 @@
+                      <help>Size of a single log file that triggers rotation</help>
+                      <valueHelp>
+                        <format>u32:1-1024</format>
+                        <description>Size in MB</description>
```
Please add a `(default: 10)` to the end of the description so the user knows it.

zdc replied:
Done.

c-po commented on interface-definitions/system-logs.xml.in:
```diff
@@ -0,0 +1,57 @@
+                      <help>Count of rotations before old logs will be deleted</help>
+                      <valueHelp>
+                        <format>u32:1-100</format>
+                        <description>Rotations</description>
```
Please add a (default: 10) to the end of the description so the user knows it.

zdc replied:
Done.

c-po commented on src/conf_mode/system-logs.py:
```diff
@@ -0,0 +1,76 @@
+
+def generate(logs_config):
+    # get configuration for logrotate atop
+    logrotate_atop = logs_config.get('logs', {}).get('logrotate',
```
Please use `dict_search('logs.logrotate.atop', logs_config)` instead.

zdc replied:
`render_to_string()` expect to get an iterable object after. 
So, it is possible to use `dict_search()` here, but we will need then also add another check for `None`/`dict` values, which will make the code even longer. Shall I switch to `dict_search()` despite this?

c-po replied:
Please do so - then all implementations are more or less lookalikes

zdc replied:
Replaced to `dict_search()`.

c-po commented on src/conf_mode/system-logs.py:
```diff
@@ -0,0 +1,76 @@
+    logrotate_atop_new = render_to_string('logs/logrotate/vyos-atop.tmpl',
+                                          logrotate_atop)
+    # update configuration files if this is necessary
+    if logrotate_atop_new != logrotate_atop_current:
```
Why is this delta check mandatory? I would always write this file - I see no harm in this as it does not happen 1000 times a second. Then you can simply use `render()` which writes to a a file.

zdc replied:
I think that blindly overwriting files with each config change is a bad design. If they are located in the `/run` this may be acceptable, but when they are in permanent storage this seems to be strange.
Maybe, we should even add an optional ability to the `util.write_file()` that can do this check, to avoid extra logic in config scripts.

c-po replied:
Well, system startup cleans files our of /etc to not have dead files lurking around. In my opinion always generating the required config files is a good design and we should not leave that path as we do it for every other part of the system. Also delta checks tend to just add additional problems as corner cases are missed.

zdc replied:
The checking was removed.

zdc committed:
logs: T3774: Improved logs CLI

Added description to the Atop section and information about default
values

zdc committed:
logs: T3774: Improved logs config rendering

Switched to `vyos.util.dict_search()` to keep the style common with the
rest components.
Removed config file comparison - almost the same result may be reached
by removing a configuration file with each boot, we already have such a
feature in the `vyos-router`.

zdc committed:
logs: T3774: Added new CLI item

Added the ability to control the `/var/log/messages` rotation.
Renamed the option `maxsize` to `max-size`.

c-po commented on data/templates/logs/logrotate/vyos-atop.tmpl:
```diff
@@ -0,0 +1,20 @@
+    daily
+    dateext
+    dateformat _%Y-%m-%d_%H-%M-%S
+    maxsize {{ max_size|default('10') }}M
```
As there is a `<defaultValue>` specified on the XML implementation the `default()` statement is superfluous.

zdc replied:
Fixed.

c-po commented on data/templates/logs/logrotate/vyos-atop.tmpl:
```diff
@@ -0,0 +1,20 @@
+    nocompress
+    nocreate
+    nomail
+    rotate {{ rotate|default('10') }}
```
As there is a `<defaultValue>` specified on the XML implementation the `default()` statement is superfluous.

c-po commented on data/templates/logs/logrotate/vyos-rsyslog.tmpl:
```diff
@@ -0,0 +1,13 @@
+    missingok
+    nomail
+    notifempty
+    rotate {{ rotate|default('10') }}
```
As there is a `<defaultValue>` specified on the XML implementation the `default()` statement is superfluous.

c-po commented on data/templates/logs/logrotate/vyos-rsyslog.tmpl:
```diff
@@ -0,0 +1,13 @@
+    nomail
+    notifempty
+    rotate {{ rotate|default('10') }}
+    size {{ max_size|default('1') }}M
```
As there is a `<defaultValue>` specified on the XML implementation the `default()` statement is superfluous.

c-po commented on src/conf_mode/system-logs.py:
```diff
@@ -0,0 +1,83 @@
+        logrotate_atop = {}
+    # generate new config file for atop
+    syslog.debug('Adding logrotate config for atop')
+    render(logrotate_atop_file, 'logs/logrotate/vyos-atop.tmpl', logrotate_atop)
```
The common design always passes `logs_config` to the `render()` function with very very very few exceptions.

The code path:
```python
    # provide an empty dictionary if there is no config
    if not logrotate_atop:
        logrotate_atop = {}
```

is superfluous as this can and should be handled in the Jinja2 template.

c-po commented on src/conf_mode/system-logs.py:
```diff
@@ -0,0 +1,83 @@
+    # get configuration for logrotate rsyslog
+    logrotate_rsyslog = dict_search('logs.logrotate.messages', logs_config)
+    # provide an empty dictionary if there is no config
+    if not logrotate_rsyslog:
```
The common design always passes `logs_config` to the `render()` function with very very very few exceptions.

The code path:
```python
    # provide an empty dictionary if there is no config
    if not logrotate_rsyslog:
        logrotate_rsyslog= {}
```

is superfluous as this can and should be handled in the Jinja2 template.

zdc replied:
Possible if we pass the whole config received via `conf.get_config_dict` to a template. There will always be a dictionary in an output. We pass here to a template only a part of the dictionary that is necessary for this template and may not exist.
In this case, I wanted to keep templates simple. Should I move this logic from Python to Jinja?

c-po replied:
Please do so - then it is consistent wirh all other implementations.

zdc replied:
Was resolved by properly applying default values. Because of them, there always will be a dictionary with values.

c-po commented on src/conf_mode/system-logs.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def apply(logs_config):
+    # No further actions needed
```
No service restart required?

zdc replied:
Not in this case. There is no service at all. Logrotate is simply started by cron/timer periodically.
More actions will be here after adding the control for `systemd-journald`.

zdc committed:
logs: T3774: Optimization for logrotate configs

* Added proper handling of default values from CLI.
* Replaced rsyslog restart postrotate action to native `rsyslog-rotate`
script.
* Removed unnecessary checks for `None` instead `dict` - with
default values the situation becomes impossible.
* Fixed default value from 10 to 1 in the rsyslog CLI.
